test/Microsoft.ML.Tests/Scenarios/Api/AspirationalExamples.cs

```diff
@@ -60,5 +60,98 @@ public void FirstExperienceWithML()
+                .Append(row => row.Label.PredictWithDecisionTrees(row.Features))
+                // shoul it be BagVectorize instead of KeyToValue?
+                .Append(row => row.PredictedLabel.KeyToValue());
+
+            var model = pipeline.Fit(data);
+
+            var predictions = model.Transform(dataReader.Read(testDataPath));
+            var evaluator = new MultiClassEvaluator(env);
+            var metrics = evaluator.Evaluate(predictions);
+        }
+
+        [Fact]
+        public void TwitterSentimentAnalysis()
+        {
+            var env = new TlcEnvironment(new SysRandom(0), verbose: true);
+            var dataPath = "wikipedia-detox-250-line-data.tsv";
+            // Load the data into the system.
+            var data = TextLoader.CreateReader(env, ctx => (
+                   Label: ctx.LoadFloat(0),
+                   Text: ctx.LoadText(1)),
+                   dataPath, hasHeader: true).Read(dataPath);
+
+            var preprocess = data.Schema.MakeEstimator(row => (
+                Label: row.Label,
+                // Concatenate all features into a vector.
+                Features: row.Text.TextFeaturizer()));
+
+            var pipeline = preprocess.
+                Append(row => row.Label.TrainLinearClassification(row.Features));
+
+            var (trainData, testData) = CrossValidator.TrainTestSplit(env, data: data, trainFraction: 0.7);
+            var model = pipeline.Fit(trainData);
+            var predictions = model.Transform(testData);
+            var evaluator = new BinaryClassifierEvaluator(env);
+            var metrics = evaluator.Evaluate(predictions);
+        }
+
+        [Fact]
+        public void TwentyNewsGroups()
+        {
+            var env = new TlcEnvironment(new SysRandom(0), verbose: true);
+            var dataPath = "20newsGroups.txt";
+            // Load the data into the system.
+            var data = TextLoader.CreateReader(env, ctx => (
+                   Label: ctx.LoadText(1),
+                   Subject: ctx.LoadText(1),
+                   Content: ctx.LoadText(2)),
+                   dataPath, hasHeader: true).Read(dataPath);
+
+            var preprocess = data.Schema.MakeEstimator(row => (
+                // Convert string label to key.
+                Label: row.Label.Dictionarize(),
+                // Concatenate all features into a vector.
+                Features: row.Subject.Concat(row.Content).TextFeaturizer()));
+
+            var pipeline = preprocess.
+                Append(row => row.Label.TrainSDCAClassifier(row.Features)).
+                Append(row => row.PredictedLabel.KeyToValue());
+
+            var (trainData, testData) = CrossValidator.TrainTestSplit(env, data: data, trainFraction: 0.8);
+            var model = pipeline.Fit(trainData);
+
+            var predictions = model.Transform(testData);
+            var evaluator = new MultiClassEvaluator(env);
+            var metrics = evaluator.Evaluate(predictions);
```

Comment:
This is insufficient. You will have to tell it *what* you want to evaluate -- that is, you will need to have some sort of delegate to select both the *label* and the *prediction* that you have done here. In fact the fact that you've attempted to do without this has hidden a nasty flaw: your `preidctions` was the result of a model whose last estimator was this:

```csharp
Append(row => row.PredictedLabel.KeyToValue());
```

You have only the predicted label, since you got rid of everything else with this simple projection.

One of the consequences of having these tuple-shape items is that we will need to use delegates, I think, to index into them. So there will be a delegate in evaluate to select the label, the predicted label, and probably the probabilities as well.

Reply:
unless it's a non-pigsty evaluator, where it can just use string names for columns... This would be `evaluator.Evaluate(predictions.DynamicData)` 

---
In reply to: [215062579](https://github.com/dotnet/machinelearning/pull/809#discussion_r215062579) [](ancestors = 215062579)

Reply:
As discussed in person, let's do
`MultiClassEvaluator.Evaluate(predictions, ctx=>predictions.label,  ctx => predictions.prediction)`, and make sure 'label' and 'prediction' are available in line 147

---
In reply to: [215062579](https://github.com/dotnet/machinelearning/pull/809#discussion_r215062579) [](ancestors = 215062579)